### ISSUE TITLE
Regression of issue 896

### ISSUE THREAD

**Issue by wulmer** [conda/conda#2151]
Hi,

since commit bc9241606ef324f1b926f077fe3940d020d3f013 there is a regression on the previous issue https://github.com/conda/conda/issues/896 where running conda takes endlessly on Windows when Anaconda is installed in a directory not writable by the normal user.

Can anyone explain why the original fix 495ea74fd95bd57a591d907a64c771260a1bdcb7 was reverted a year later in bc9241606ef324f1b926f077fe3940d020d3f013 ?

This makes Anaconda totally unusable in a Windows installation *For all users*.

Wolfgang

**Comment by msarahan** [comment]
Ping @groutr - looks like you merged this at https://github.com/conda/conda/commit/9dd141c2d534529bc9e491fcdcb7b0518e9418ba

I think we need a comment in the code at the very least, and ideally a regression test to make sure this doesn't happen again.

**Comment by kalefranz** [comment]
Also tagging @pelson here since some of his commits look to be involved.

@wulmer I don't think the original fix was intentionally reverted, but rather @pelson stumbled into the same solution when addressing #1734.

**Comment by pelson** [comment]
Whilst the solution proposed in bc92416 is correct, it is also clearly problematic on Windows.

@wulmer - could you try timing Tempfile usage on a directory you don't have access to on Windows? My assumption is that the performance is disastrous, and fundamentally is a Tempfile issue. Assuming that to be the case, do you know if that is documented somewhere or even an issue on the Python tracker?

Even so, we are going to have to find a solution to this. It might make sense to have a different code path for Windows where we inspect the filesystem for writability, rather than asking forgiveness if we can't write to the desired location.

**Comment by wulmer** [comment]
@pelson Sure, here is the benchmark (Windows 7, SSD drive, Intel Core i5, 2.6 GHz):

    In [7]: %%timeit
       ...: try:
       ...:     tempfile.NamedTemporaryFile(dir='C:/Program Files/')
       ...: except:
       ...:     pass
       ...:
    1 loop, best of 3: 8.26 s per loop

An simple and stupid solution for keeping `tempfile` while improving the performance is to monkey-patch `os.TMP_MAX` temporarily before calling `tempfile`'s method.
This constant sets the number of retries for temporary file creation (see [source code of `tempfile`](https://hg.python.org/cpython/file/tip/Lib/tempfile.py)).

**Comment by pelson** [comment]
Amazing! Thanks for responding so quickly @wulmer. We could monkey-patch TMP_MAX, but I'm just as inclined to try inspecting the permissions. Presumably http://stackoverflow.com/a/2113511/741316 gives you a sensible answer (including the ```os.X_OK``` mentioned in the comments).

I also think this would be a useful thing to be able to highlight on the CPython issue tracker if it hasn't already been put on there - who knew that tempfile could be so poor when you don't have write permission!

**Comment by wulmer** [comment]
I tried several methods of inspecting permissions, but nothing seems to work reliably. For example `os.access` should be able to check at least if I don't have access rights on a folder, but even this doesn't work:
```python
>>> os.access('C:/Windows', os.W_OK)
True
>>> open('C:/Windows/file.txt', 'w')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IOError: [Errno 13] Permission denied: 'C:/Windows/file.txt'
```
There is a [note in the documentation](https://docs.python.org/dev/library/os.html?highlight=xattr#os.access) which says:
> I/O operations may fail even when access() indicates that they would succeed, particularly for operations on network filesystems which may have permissions semantics beyond the usual POSIX permission-bit model. 

And other people have reported of similar problems: [\[1\]](http://code.activestate.com/lists/python-list/617035/) and [\[2\]](https://github.com/lucadelu/pyModis/issues/34).

The only reliable method seems to be to open a file for writing in that directory (i.e. what `tempfile` does). However using `tempfile` for that (with its many retries in case of failure) is not necessary in my opinion.

Can't we try to create a file using standard `open` method like in 495ea74fd95bd57a591d907a64c771260a1bdcb7 with the only difference that we are using a unique file name derived from the current process ID, something like `".conda-try-write-%d" % os.getpid()`? A first simple implementation of this idea is under https://github.com/conda/conda/pull/2301.

**Comment by groutr** [comment]
Wow.  That's really good to know about tempfile.  I certainly wasn't aware of it.  I'm +1 for using a unique filename though as @wulmer suggests.

**Comment by wulmer** [comment]
OK. Solved in PR #2301.
